### ISSUE TITLE
Question: possibilty of byte level corruption (not atomicity violation) when appending to an existing file in various file system and configurations

### ISSUE THREAD

**Issue by rotemdan** [madthanu/alice#5]
I apologize if this isn't the best place to ask this but it seemed to be the best I could find on the web. Unfortunately this is a subject that very few people are knowledgeable enough to answer with precision. I would be extremely thankful if anyone could point me to an answer for this!

I have read and consulted with the USENIX paper many times and I still couldn't confidently extract an answer. This is important for a transactional storage engine I'm developing.

My question is:

Given an existing file (assume its initial existence and data are safely persisted on physical media), that's accessed strictly from a single process, and opened in read/append only mode. Now this file is appended with an arbitrary length of data (possibly larger than the sector/block size) in a single thread of operation. Consider these two cases:

1. _One `append` operation_ i.e. only a single call to OS `write`, or,
2. _Multiple consecutive `append` operations_.

In what file-system configurations it is possible that a crash scenario would lead to the file containing individual bytes that do not correspond to the data that was intended to be written? (regardless of whether the write has completed in its entirety, i.e. it is completely acceptable that the content is truncated or not written at all). I would also be very thankful to know the particular case of the NTFS file system (not mentioned in the article).

There is one sentence that seems to address an issue related to this, but I'm not sure how to interpret it:

> A similar special case is appending to files that have been opened with O_TRUNC. Even with delayed allocation, successive appends to the same file are persisted in order.

Does this mean that if an existing file with existing data is opened (i.e. not opened with `O_TRUNC`), there is a possibility that successive appends might be persisted in an arbitrary order, such that the resulting file might contain areas filled with `0`s or arbitrary data?

_Edit_: To rephrase this in a simpler way, can the **size** of the file be trusted in this scenario? such that whatever exists up to that size is ensured to be data that was explicitly passed to be written, i.e. the file is guaranteed not to contain areas of `0`s or arbitrary bytes?

**Comment by madthanu** [comment]
Unfortunately, you will not find an exact answer here, either :).

We (Alice, OSDI'14) did not study NTFS, and I can't answer to that. With ext3/ext4, it used to be that you can trust the file size unless you mounted the file system with data=writeback mode. But, even if you did not mount it in this mode, in some versions of ext4, you cannot fully trust the file size -- the appended portion might contain zeros (but not garbage); I believe this was a file system bug that was then fixed in later versions of ext4, but I am not sure.

Ultimately, even if you figure out that you can trust the file size for a particular file-system configuration, the file system might start behaving differently in a future version. Remember that file systems don't really *guarantee* anything with respect to your question. I'll be really careful assuming something.

**Comment by rotemdan** [comment]
Thank you for answering so quickly, I did not expect that. I have been dealing with this issue for almost a year now. The closest answer I found (apart from the USENIX article) so far was [this StackOverflow question](https://stackoverflow.com/questions/14324151/can-file-size-be-used-to-detect-a-partial-append), which unfortunately didn't get a definitive answer.

The fact you mention that even in ordered mode there used to be a chance for corruption is a bit worrying.

The reason this is so important is that I'm trying to develop a storage engine that both avoids checksumming of data and allows to be used with delayed allocation, optionally without `fsync` calls, and without the risk of corruption (loss of data is acceptable). I thought I could just use the file size as a metric to whether a transaction has been fully committed (or needs to be rolled-back). I hoped this would at least be reliable in some file system configurations (the reasons are for performance, mainly).

It seems like in this light, I'll need to think of some additional alternatives, or just introduce some form of checksumming.
